rngd_qrypt.c

```diff
@@ -0,0 +1,317 @@
+	size_t realsize = size * nmemb;
+	char *ptr = realloc(buf->response, buf->size + realsize + 1);
+
+	if (!ptr) {
+		return 0;
+	}
+	buf->response = ptr;
+	memcpy(&(buf->response[buf->size]), data, realsize);
+	buf->size += realsize;
+	buf->response[buf->size] = 0;
+	return realsize;
+}
+
+static void *refill_task(void *data __attribute__((unused)))
+{
+	CURL *curl;
+	struct curl_slist *list = NULL;
+	CURLcode res;
+	long response_code;
+	struct body_buffer response_data;
+	pthread_mutex_lock(&ent_lock);
+	refilling = true;
+	pthread_cond_signal(&ent_cond);
+	pthread_mutex_unlock(&ent_lock);
+
+	response_data.response = NULL;
+	response_data.size = 0;
+	curl = curl_easy_init();
+	if (!curl) {
+		message_entsrc(my_ent_src, LOG_DAEMON|LOG_INFO, "Unable to init curl\n");
+		goto out;
+	}
+	curl_easy_setopt(curl, CURLOPT_URL, QRYPT_URL);
+	curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
+	curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
+	curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);
+
+	list = curl_slist_append(list, "Accept: application/json");
+	list = curl_slist_append(list, bearer); 
+
+	curl_easy_setopt(curl, CURLOPT_HTTPHEADER, list);
+	
+	res = curl_easy_perform(curl);
+	if (res != CURLE_OK) {
+		message_entsrc(my_ent_src, LOG_DAEMON|LOG_INFO, "Failed to send curl: %d\n", res);
+		goto out;
+	}
+
+	curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &response_code);
+	if (response_code != 200) {
+		message_entsrc(my_ent_src, LOG_DAEMON|LOG_INFO, "qrypt server responds not ok: %lu\n", response_code);
+		goto out;
+	}
+	extract_and_refill_entropy(&response_data);
+	
+out:
+	if (curl)
+		curl_easy_cleanup(curl);
+	if (response_data.response)
+		free(response_data.response);
+	pthread_mutex_lock(&ent_lock);
+	refilling = false;
+	pthread_cond_signal(&ent_cond);
+	pthread_mutex_unlock(&ent_lock);
+	return NULL;
+}
+
+static void refill_ent_buffer()
+{
+	pthread_t tid;
+	pthread_attr_t tattr;
+	pthread_attr_init(&tattr);
+	pthread_attr_setdetachstate(&tattr, PTHREAD_CREATE_DETACHED);
+	pthread_mutex_lock(&ent_lock);
+	if (refilling) {
+		pthread_cond_wait(&ent_cond, &ent_lock);
+		if (refilling) {
+			message_entsrc(my_ent_src, LOG_DAEMON|LOG_INFO, "refilling taking a long time, aborting\n");
+		}
+		goto out;
+	}
+	pthread_create(&tid, &tattr, refill_task, NULL);
+	pthread_cond_wait(&ent_cond, &ent_lock);
+out:
+	pthread_mutex_unlock(&ent_lock);
+}
+
+int xread_qrypt(void *buf, size_t size, struct rng *ent_src)
+{
+	size_t new_avail = 0;
+	bool satisfied = false;
+	pthread_mutex_lock(&ent_lock);
+	if (avail_ent >= size) {
+		memcpy(buf, &entropy_buffer[ent_idx], size);
+		avail_ent -= size;
+		ent_idx += size;
+		new_avail = avail_ent;
+		satisfied = true;
+	}
+	pthread_mutex_unlock(&ent_lock);
+	if (new_avail <= REFILL_THRESH) {
+		refill_ent_buffer();
+	}
+	return satisfied ? 0 : -1;
+}
+
+
+/*
+ * Init QRYPT 
+ */
+int init_qrypt_entropy_source(struct rng *ent_src)
```

Comment:
The token is assumed to be 2048 bytes exactly right now. It could be good to make this more future-proof and allowing this to grow later if necessary. We should assume an upper bound of 16K for the entire header, as this tends to be a typical HTTP server limitation. How about using `fgets` instead of `fread` to read the token line from the file, then doing `token[strcspn(token, "\r\n")] = 0` to remove any possible carriage returns if the user saved them into the file, and then copying this into `bearer` (could potentially use `asnprintf` instead of a statically allocated buffer for `bearer`).

Reply:
fgets still nees to know an uppper limit to the amount of data it can read.  If the goal is to future proof the size of the bearer token, the best solution is likely to fstat the file, allocate a buffer commiserate with the size returned from that call, and then we can use fgets or read as needed